### PR TITLE
add optionality of time based score, punishment score and buzzer twice

### PR DESCRIPTION
**Pull request by yuvalvaknin**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -33,6 +33,7 @@
     "class-validator": "^0.14.1",
     "fastest-levenshtein": "^1.0.16",
     "ioredis": "^5.4.1",
+    "nest": "^0.1.6",
```
Comment (orenhta) [inline package.json]: why?

File: src/dto/create-game-request.dto.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { IsBoolean, IsNotEmpty } from "class-validator";
+
+export class CreateGameRequestDto {
+    @IsNotEmpty()
```
Comment (orenhta) [inline src/dto/create-game-request.dto.ts]: are those `@IsNotEmpty()` needed?

File: src/dto/create-game-request.dto.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    @IsBoolean()
+    isTimeBasedScore: boolean;
+
+    @IsNotEmpty()
+    @IsBoolean()
+    isPunishmentScoreAllowed: boolean;
+
+    @IsNotEmpty()
+    @IsBoolean()
+    allowBuzzerTwice: boolean;
```
Comment (orenhta) [inline src/dto/create-game-request.dto.ts]: use the same naming convention. `isBuzzerTwiceAllowed`

File: src/enums/emitted-events.enum.ts [hunk 4]
```diff
@@ -1,6 +1,7 @@
 export enum EmittedEvent {
   BUZZER_GRANTED = 'buzzer-granted',
   BUZZER_REVOKED = 'buzzer-revoked',
+  BUZZER_CONFLICT='buzzer-conflict',
```
Comment (orenhta) [inline src/enums/emitted-events.enum.ts]: spaces

File: src/modules/game-manager/game-manager.gateway.ts [hunk 5]
```diff
@@ -42,10 +43,11 @@ export class GameManagerGateway implements OnGatewayDisconnect {
 
   @SubscribeMessage('create-game')
   async handleCreateGame(
+    payload: CreateGameRequestDto,
```
Comment (orenhta) [inline src/modules/game-manager/game-manager.gateway.ts:47]: `@MessageBody(defaultValidationPipe) payload: CreateGameRequestDto,`

File: src/modules/game-manager/game-manager.service.ts [hunk 6]
```diff
@@ -55,6 +56,9 @@ export class GameManagerService {
       totalRounds,
       songs,
       roundData: {},
+      isTimeBasedScore: createGameRequestDto.isTimeBasedScore,
+      isPunishmentScoreAllowed: createGameRequestDto.isPunishmentScoreAllowed,
+      allowBuzzerTwice: createGameRequestDto.allowBuzzerTwice
```
Comment (orenhta) [inline src/modules/game-manager/game-manager.service.ts]: put those in a `gameSettings` property

Comment (orenhta) [reply src/modules/game-manager/game-manager.service.ts]: and also in the type

File: src/modules/game-client/game-client.service.ts [hunk 7]
```diff
@@ -55,79 +56,95 @@ export class GameClientService {
...
-      await this.gameStateRepository.getGameState<GameStatus.ROUND_IN_PROGRESS>(
-        gameId,
-      );
+        await this.gameStateRepository.getGameState<GameStatus.ROUND_IN_PROGRESS>(
+          gameId,
+        );
+    const player: Player = gameState.gamePlayers[socketId];
 
-    const player = gameState.gamePlayers[socketId];
+    if (socketId === gameState.roundData.buzzersGranted[gameState.roundData.buzzersGranted.length]){
```
Comment (orenhta) [inline src/modules/game-client/game-client.service.ts:64]: shouldn’t you implement this logic only if allow buzzer twice is not allowed?

Comment (orenhta) [inline src/modules/game-client/game-client.service.ts]: why are you checking if this is a buzzer twice here?

shouldn’t you also check if its allowed in the settings

Comment (orenhta) [reply src/modules/game-client/game-client.service.ts]: didn’t understand this logic brother in this function.

also why is the `EmittedEvent.BUZZER_CONFLICT` necessary?

**Comment by orenhta** [comment]
brother, pull from main to get the latest changes.

File: src/dto/create-game-request.dto.ts (hunk 2 above)
Comment (yuvalvaknin) [reply src/dto/create-game-request.dto.ts]: yes I want them all as a must parameters

File: package.json (hunk 1 above)
Comment (yuvalvaknin) [reply package.json]: No reason mate that file shouldnt be included in the commit.
My bad

File: src/dto/create-game-request.dto.ts (hunk 3 above)
Comment (yuvalvaknin) [reply src/dto/create-game-request.dto.ts]: 9b8869398cc1385451e0f1ecab8fd707af4b7456

File: src/enums/emitted-events.enum.ts (hunk 4 above)
Comment (yuvalvaknin) [reply src/enums/emitted-events.enum.ts]: 9b8869398cc1385451e0f1ecab8fd707af4b7456

File: src/modules/game-manager/game-manager.gateway.ts (hunk 5 above)
Comment (yuvalvaknin) [reply src/modules/game-manager/game-manager.gateway.ts:47]: 9b8869398cc1385451e0f1ecab8fd707af4b7456

File: package.json (hunk 1 above)
Comment (yuvalvaknin) [reply package.json]: 78e2de8a456475b200297f847e9ca29e581b6162

File: src/dto/create-game-request.dto.ts (hunk 2 above)
Comment (orenhta) [reply src/dto/create-game-request.dto.ts]: @IsBoolean() does this

**Pull request merged**